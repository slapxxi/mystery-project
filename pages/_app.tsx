import AppContainer from '@self/components/AppContainer';
import { appWithTranslation } from '@self/i18n';
import { AppProps } from '@self/lib/types';
import App from 'next/app';
import React from 'react';

class MyApp extends App {
  static async getInitialProps(context: any) {
    let { Component, ctx } = context;
    let { req } = ctx;
    let pageProps: AppProps = { user: null };

    if (req) {
      let { session } = req;

      if (session && session.decodedToken) {
        pageProps = { ...pageProps, user: session.decodedToken };
      }
    }

    if (Component.getInitialProps) {
      pageProps = {
        ...pageProps,
        ...(await Component.getInitialProps(ctx)),
      };
    }

    return { pageProps };
  }

  render() {
    let { Component, pageProps } = this.props;

    return (
      <AppContainer user={pageProps.user} posts={pageProps.posts}>
        <Component {...pageProps} />
      </AppContainer>
    );
  }
}

export default appWithTranslation(MyApp);
