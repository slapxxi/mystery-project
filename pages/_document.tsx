import Document, {
  Head,
  Html,
  Main,
  NextDocumentContext,
  NextScript,
} from 'next/document';

class MyDocument extends Document {
  static async getInitialProps(ctx: NextDocumentContext) {
    let initialProps = await Document.getInitialProps(ctx);
    return { ...initialProps };
  }

  render() {
    return (
      <Html>
        <Head />
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

export default MyDocument;
