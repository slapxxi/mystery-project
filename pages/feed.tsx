/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import PostsGrid from '@self/components/PostsGrid';
import Spinner from '@self/components/Spinner';
import { Link, withTranslation } from '@self/i18n';
import getUser from '@self/lib/getUser';
import useAuth from '@self/lib/hooks/useAuth';
import routes from '@self/lib/routes';
import fetchPosts from '@self/lib/services/fetchPosts';
import fetchSubscriptions from '@self/lib/services/fetchSubscriptions';
import {
  AuthUser,
  Maybe,
  PageContext,
  PagePropsWithTranslation,
  Post,
  Subscription,
} from '@self/lib/types';
import { useMachine } from '@xstate/react';
import { assign, Machine } from 'xstate';

interface Props extends PagePropsWithTranslation<'common' | 'header'> {
  posts: Post[];
  subscriptions?: Subscription[];
}

interface Context {
  user: Maybe<AuthUser>;
  subscriptions: Maybe<Subscription[]>;
  error: Maybe<Error>;
}

let pageMachine = Machine<Context>({
  id: 'feed-page',
  initial: 'idle',
  context: {
    user: null,
    subscriptions: null,
    error: null,
  },
  states: {
    idle: {
      on: {
        '': [
          { target: 'success', cond: (context) => !!context.subscriptions },
          { target: 'loading', cond: (context) => !!context.user },
        ],
      },
    },
    loading: {
      invoke: {
        src: 'fetchSubscriptions',
        onDone: {
          target: 'success',
          actions: assign({ subscriptions: (_: Context, event: any) => event.data }),
        },
        onError: {
          target: 'error',
          actions: assign({ error: (_: Context, event: any) => event.data }),
        },
      },
    },
    error: {
      on: { RETRY: 'loading' },
    },
    success: { type: 'final' },
  },
});

function FeedPage(props: Props) {
  let { posts, subscriptions, t } = props;
  let [authState] = useAuth();
  let [pageState, send] = useMachine(pageMachine, {
    context: { user: authState.context.user, subscriptions },
    services: {
      fetchSubscriptions: (context: Context) => fetchSubscriptions(context.user),
    },
  });

  return (
    <div>
      {authState.matches('auth') && (
        <Link href={routes.posts.new.url}>
          <a href="">{t('create post')}</a>
        </Link>
      )}
      {pageState.matches('loading') ? (
        <Spinner></Spinner>
      ) : pageState.matches('success') ? (
        <>
          <h2>Following</h2>
          <PostsGrid posts={pageState.context.subscriptions}></PostsGrid>
        </>
      ) : pageState.matches('error') ? (
        <>
          <span css={errorStyles}>{pageState.context.error.message}</span>
          <button onClick={() => send('RETRY')}>Retry</button>
        </>
      ) : null}
      <h2>Popular</h2>
      <PostsGrid posts={posts}></PostsGrid>
    </div>
  );
}

FeedPage.getInitialProps = async (context: PageContext) => {
  let user = getUser(context);
  let posts: Post[] = [];
  let subscriptions: Maybe<Subscription[]> = null;

  if (user) {
    subscriptions = await fetchSubscriptions(user);
  }

  posts = await fetchPosts();

  return { namespacesRequired: ['common', 'header'], posts, subscriptions };
};

let errorStyles = css`
  display: inline-block;
  background: tomato;
  padding: 10px;
  border-radius: 4px;
  color: hsl(0, 30%, 90%);
`;

// @ts-ignore
export default withTranslation('common')(FeedPage);
