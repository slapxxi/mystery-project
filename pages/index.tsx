/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import '@self/styles/normalize.css';

function IndexPage() {
  return (
    <div
      css={css`
        width: 960px;
        margin: 0 auto;
      `}
    >
      <h1>Index Page</h1>
    </div>
  );
}

export default IndexPage;
