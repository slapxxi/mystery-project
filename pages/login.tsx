/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import GithubIcon from '@self/components/icons/GithubIcon';
import Spinner from '@self/components/Spinner';
import { Router, withTranslation } from '@self/i18n';
import useAuth from '@self/lib/hooks/useAuth';
import isServer from '@self/lib/isServer';
import redirectTo from '@self/lib/redirectTo';
import { AppTheme, PageContext, PagePropsWithTranslation } from '@self/lib/types';
import userAuthenticated from '@self/lib/universal/userAuthenticated';
import 'firebase/auth';

interface Props extends PagePropsWithTranslation<'common'> {}

function LoginPage(props: Props) {
  let { t } = props;
  let [auth, actions] = useAuth();

  if (auth.matches('auth')) {
    Router.push('/feed');

    return <div>Successfully Signed In</div>;
  }

  async function handleClick() {
    actions.signIn();
  }

  if (auth.matches('signingIn')) {
    return (
      <div>
        <Spinner></Spinner>
      </div>
    );
  }

  return (
    <div>
      <h1>{t('login')}</h1>
      <button
        data-testid="signin"
        onClick={handleClick}
        // disabled={auth.status === 'pending'}
        css={(theme: AppTheme) => css`
          display: inline-flex;
          align-items: center;
          border: 0;
          background-color: ${theme.type === 'light' ? 'black' : 'white'};
          color: ${theme.type === 'light' ? 'white' : 'black'};
          border-radius: 4px;
          padding: 10px;

          :hover {
            background-color: ${theme.type === 'light'
              ? 'hsl(0, 10%, 10%)'
              : 'hsl(0, 10%, 90%)'};
          }
        `}
      >
        <GithubIcon
          width={24}
          css={(theme: AppTheme) => css`
            fill: ${theme.type === 'light' ? 'white' : 'black'};
            margin-right: 10px;
            margin-top: -4px;
          `}
        ></GithubIcon>
        {t('sign in with github')}
      </button>
    </div>
  );
}

LoginPage.getInitialProps = async (context: PageContext) => {
  if (isServer(context) && userAuthenticated(context)) {
    redirectTo(context, '/');
  }

  return { namespacesRequired: ['common', 'header'] };
};

export default withTranslation('common')(LoginPage);
