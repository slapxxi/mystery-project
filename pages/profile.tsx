/** @jsx jsx */
import { jsx } from '@emotion/core';
import { Router, withTranslation } from '@self/i18n';
import useAuth from '@self/lib/hooks/useAuth';
import isServer from '@self/lib/isServer';
import redirectTo from '@self/lib/redirectTo';
import { PageContext, PagePropsWithTranslation } from '@self/lib/types';
import userAuthenticated from '@self/lib/universal/userAuthenticated';

interface Props extends PagePropsWithTranslation<'common'> {}

function ProfilePage(props: Props) {
  let { t } = props;
  let [auth, authActions] = useAuth();

  if (!auth.user) {
    Router.push('/');
  }

  function handleClick() {
    authActions.requestSignOut();
  }

  return (
    <div>
      <h1>{t('profile')}</h1>
      <button onClick={handleClick}>Sign Out</button>
    </div>
  );
}

ProfilePage.getInitialProps = async (context: PageContext) => {
  if (isServer(context) && !userAuthenticated(context)) {
    redirectTo(context, '/');
  }

  return { namespacesRequired: ['common', 'header'] };
};

export default withTranslation('common')(ProfilePage);
