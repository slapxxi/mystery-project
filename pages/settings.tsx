/** @jsx jsx */
import { jsx } from '@emotion/core';

function SettingsPage() {
  return (
    <div>
      <h1>Settings Page</h1>
    </div>
  );
}

export default SettingsPage;
