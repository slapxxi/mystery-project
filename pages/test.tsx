import { useMachine } from '@xstate/react';
import { Machine } from 'xstate';

interface Context {
  ts: number;
}

interface State {
  states: {
    idle: {};
    activating: {};
    active: {};
  };
}

type Events = { type: 'INIT' };

let testMachine = Machine<Context, State, Events>({
  id: 'test-machine',
  initial: 'idle',
  context: { ts: null },
  states: {
    idle: {
      entry: ['log'],
      activities: ['beeping'],
      on: {
        INIT: { target: 'activating', cond: 'requirementsMet' },
      },
    },
    activating: { invoke: { src: 'fetchData', onDone: 'active' } },
    active: { type: 'final' },
  },
});

function TestingPage(_props: any) {
  let [state, send] = useMachine(testMachine, {
    context: { ts: Date.now() },
    guards: {
      requirementsMet: (context, event) => Date.now() - context.ts > 3000,
    },
    actions: { log: console.log },
    activities: {
      beeping: () => {
        let id = setInterval(() => console.log('beep'), 1000);
        return () => clearInterval(id);
      },
    },
    services: {
      fetchData,
    },
  });

  console.log(state.event);
  console.log(state.value);

  return (
    <div>
      <h1>{state.matches('active') ? 'Active' : 'Inactive'}</h1>
      <button onClick={() => send('INIT')}>Activate</button>
    </div>
  );
}

function fetchData() {
  return new Promise((res) => {
    setTimeout(() => res(), 3000);
  });
}

export default TestingPage;
