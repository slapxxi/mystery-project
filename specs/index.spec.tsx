import { IndexPage } from '@self/pages/index';
import fs from 'fs';
import path from 'path';
import React from 'react';
import { render } from 'react-testing-library';

let translation = JSON.parse(
  fs.readFileSync(
    path.join(__dirname, '..', 'static', 'locales', 'en', 'common.json'),
    'utf-8'
  )
);

let translate = jest.fn((key) => translation[key]);

it('renders', () => {
  // @ts-ignore
  let { getByTestId } = render(<IndexPage t={translate} user={null} />);
  let pageTitle = getByTestId('title');

  expect(pageTitle.textContent).toEqual('Home');
});
