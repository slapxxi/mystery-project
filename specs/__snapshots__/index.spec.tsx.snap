// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <h1>
    Index Page
  </h1>
</div>
`;
